v3_async_wip/tests/test_iothub_mqtt_client.py

```diff
@@ -2478,35 +2699,189 @@ async def test_message_property_defaults(self, client):
+        mock_factory = mocker.patch.object(models.Message, "create_from_properties_dict")
+
+        def fail_once(*args, **kwargs):
+            mock_factory.side_effect = original_factory
+            raise arbitrary_exception
+
+        mock_factory.side_effect = fail_once
+
+        # Load the MQTTMessages
+        await client._mqtt_client._incoming_filtered_messages[sub_topic].put(mqtt_msg1)
+        await client._mqtt_client._incoming_filtered_messages[sub_topic].put(mqtt_msg2)
+
+        # The Message is derived from the second MQTTMessage instead of the first because the
+        # first failed, the error was suppressed, and the MQTTMessage discarded
+        msg = await client.incoming_input_messages.__anext__()
+        assert msg.payload == payload2
+        assert payload2 != payload1
+        assert mock_factory.call_count == 2
+
+    @pytest.mark.skip(reason="Currently can't figure out how to mock a generator correctly")
+    @pytest.mark.it("Can be cancelled while waiting for an MQTTMessage to arrive")
+    async def test_cancelled_while_waiting_for_message(self):
+        pass
+
+
+@pytest.mark.describe("IoTHubMQTTClient - PROPERTY: .incoming_direct_method_requests")
+class TestIoTHubMQTTClientIncomingDirectMethodRequests:
+    @pytest.mark.it("Is an AsyncGenerator maintained as a read-only property")
+    def test_property(self, client):
+        assert isinstance(client.incoming_direct_method_requests, typing.AsyncGenerator)
+        with pytest.raises(AttributeError):
+            client.incoming_direct_method_requests = 12
+
+    @pytest.mark.it(
+        "Yields a DirectMethodRequest whenever the MQTTClient receives an MQTTMessage on the incoming direct method request topic"
+    )
+    async def test_yields_direct_(self, client):
+        generic_topic = mqtt_topic.get_direct_method_request_topic_for_subscribe()
+
+        # Create MQTTMessages
+        mreq1_name = "some_method"
+        mreq1_id = "12"
+        mreq1_topic = generic_topic.rstrip("#") + "{}/?$rid={}".format(mreq1_name, mreq1_id)
+        mqtt_msg1 = mqtt.MQTTMessage(mid=1, topic=mreq1_topic.encode("utf-8"))
+        mqtt_msg1.payload = '{"json": "in", "a": {"string": "format"}}'.encode("utf-8")
+        mreq2_name = "some_other_method"
+        mreq2_id = "17"
+        mreq2_topic = generic_topic.rstrip("#") + "{}/?$rid={}".format(mreq2_name, mreq2_id)
+        mqtt_msg2 = mqtt.MQTTMessage(mid=2, topic=mreq2_topic.encode("utf-8"))
+        mqtt_msg2.payload = '{"json": "in", "a": {"different": {"string": "format"}}}'.encode(
+            "utf-8"
+        )
+
+        # Load the MQTTMessages into the MQTTClient's filtered message queue
+        await client._mqtt_client._incoming_filtered_messages[generic_topic].put(mqtt_msg1)
+        await client._mqtt_client._incoming_filtered_messages[generic_topic].put(mqtt_msg2)
+
+        # Get items from generator
+        mreq1 = await client.incoming_direct_method_requests.__anext__()
```

Comment:
will this not work ?
mreq1 = await anext(client.incoming_direct_method_requests)

Reply:
It would in most versions of Python, but not 3.7